state/execution_test.go

```diff
@@ -30,16 +32,53 @@ func TestApplyBlock(t *testing.T) {
 	state := state()
 	state.SetLogger(log.TestingLogger())
 
-	// make block
 	block := makeBlock(1, state)
 
 	err = state.ApplyBlock(types.NopEventBus{}, proxyApp.Consensus(), block, block.MakePartSet(testPartSize).Header(), types.MockMempool{})
-
 	require.Nil(t, err)
 
 	// TODO check state and mempool
 }
 
+// TestBeginBlockAbsentValidators ensures we send absent validators list.
+func TestBeginBlockAbsentValidators(t *testing.T) {
+	app := &testApp{}
+	cc := proxy.NewLocalClientCreator(app)
+	proxyApp := proxy.NewAppConns(cc, nil)
+	err := proxyApp.Start()
+	require.Nil(t, err)
+	defer proxyApp.Stop()
+
+	state := state()
+	state.SetLogger(log.TestingLogger())
+
+	// there were 2 validators
+	val1PrivKey := crypto.GenPrivKeyEd25519()
+	val2PrivKey := crypto.GenPrivKeyEd25519()
+	lastValidators := types.NewValidatorSet([]*types.Validator{
+		types.NewValidator(val1PrivKey.PubKey(), 10),
+		types.NewValidator(val2PrivKey.PubKey(), 5),
+	})
+
+	// but last commit contains only the first validator
+	prevHash := state.LastBlockID.Hash
+	prevParts := types.PartSetHeader{}
+	prevBlockID := types.BlockID{prevHash, prevParts}
+	lastCommit := &types.Commit{BlockID: prevBlockID, Precommits: []*types.Vote{
+		{ValidatorIndex: 0},
+	}}
+
+	valHash := state.Validators.Hash()
+	block, _ := types.MakeBlock(2, chainID, makeTxs(2), lastCommit,
```

Comment:
Nice test, but with all this setup, why not make a table-driven test and verify a few different situations, just loop over thel last 10 lines or so with different variables (none absent, one absent, multiple absent)?